test.js

```diff
@@ -1,95 +1,270 @@
+    t.plan(1)
+    rp({
+      method: 'GET',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      uri: 'http://localhost:' + fastify.server.address().port + '/verify',
+      json: true
+    })
+      .then(function () {
+        t.fail()
+      })
+      .catch(function (err) {
+        t.is(err.error.message, 'No Authorization was found in request.headers')
+      })
+  })
+
+  t.test('jwtVerify throws Authorization Format error', function (t) {
+    t.plan(1)
+    rp({
+      method: 'GET',
+      headers: {
+        'Content-Type': 'application/json',
+        authorization: 'Invalid TokenFormat'
+      },
+      uri: 'http://localhost:' + fastify.server.address().port + '/verify',
+      json: true
+    })
+      .then(function () {
+        t.fail()
+      })
+      .catch(function (err) {
+        t.is(err.error.message, 'Format is Authorization: Bearer [token]')
+      })
+  })
+
+  t.test('jwtSign throws payload error', function (t) {
+    t.plan(1)
+    rp({
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify({
+        notAPayload: 'sorry'
+      }),
+      uri: 'http://localhost:' + fastify.server.address().port + '/sign',
+      json: true
+    })
+      .then(function () {
+        t.fail()
+      })
+      .catch(function (err) {
+        t.is(err.error.message, 'jwtSign requires a payload')
+      })
+  })
 })
 
-test('should throw if no secret is given as option', t => {
+test('decode', function (t) {
   t.plan(1)
-  const fastify = Fastify()
-  fastify.register(jwt, err => {
-    t.is(err.message, 'missing secret')
+  var fastify = Fastify()
+  fastify.register(jwt, { secret: 'shh' }).ready(function () {
+    var token = fastify.jwt.sign({ foo: 'bar' })
+    var decoded = fastify.jwt.decode(token)
+    t.is(decoded.foo, 'bar')
   })
 })
 
-test('sign should throw if the payload is missing', t => {
-  t.plan(2)
-  const fastify = Fastify()
+test('secret as a function', function (t) {
+  t.plan(4)
+  var fastify = Fastify()
   fastify
-    .register(jwt, { secret: 'supersecret' }, t.error)
-    .after(() => {
-      try {
-        fastify.jwt.sign()
-        t.fail()
-      } catch (err) {
-        t.is(err.message, 'missing payload')
+    .register(jwt, {
+      secret: function (request, reply, callback) {
+        callback(null, 'supersecret')
       }
     })
-})
 
-test('verify should throw if the token is missing', t => {
-  t.plan(2)
-  const fastify = Fastify()
-  fastify
-    .register(jwt, { secret: 'supersecret' }, t.error)
-    .after(() => {
-      try {
-        fastify.jwt.verify()
-        t.fail()
-      } catch (err) {
-        t.is(err.message, 'missing token')
-      }
+  fastify.post('/sign', function (request, reply) {
+    reply.jwtSign(request.body.payload, function (err, token) {
+      if (err) { return reply.send(err) }
+      return reply.send({token})
+    })
+  })
```

Comment:
can you make `jwtSign()` and `jwtVerify()` return a `Promise` if no callback is passed through?

This would enable a nice async/await syntax.

Reply:
I hope I added this functionality correctly. I'll write some unit tests now to test it out.